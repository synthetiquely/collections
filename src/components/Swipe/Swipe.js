import React, { Component } from 'react';
import glamorous from 'glamorous';

const SwipeContainer = glamorous.div({
  width: '100%',
  height: '100%',
  touchAction: 'none',
});

class Swipe extends Component {
  constructor(props) {
    super(props);
    this.surface = null;
    this.direction = 'none';
    this.startX = 0;
    this.startY = 0;
    this.distX = 0;
    this.distY = 0;
    this.threshold = 150;
    this.restraint = 100;
    this.allowedTime = 200;
    this.startTime = 0;
    this.setRef = this.setRef.bind(this);
    this.onTouchStart = this.onTouchStart.bind(this);
    this.onTouchEnd = this.onTouchEnd.bind(this);
  }

  onTouchStart(e) {
    e.preventDefault();

    const touchObj = e.changedTouches[0];
    this.distX = 0;
    this.distY = 0;
    this.startX = touchObj.pageX;
    this.startY = touchObj.pageY;
    this.startTime = new Date().getTime();
    this.direction = 'none';
  }

  onTouchEnd(e) {
    e.preventDefault();

    const touchObj = e.changedTouches[0];
    this.distX = touchObj.pageX - this.startX;
    this.distY = touchObj.pageY - this.startY;

    if (Math.abs(this.distX) > Math.abs(this.distY)) {
      this.direction = this.distX < 0 ? 'left' : 'right';
    } else {
      this.direction = this.distY < 0 ? 'up' : 'down';
    }

    const elapsedTime = new Date().getTime() - this.startTime;
    if (elapsedTime <= this.allowedTime) {
      if (
        Math.abs(this.distX) >= this.threshold &&
        Math.abs(this.distY) <= this.restraint
      ) {
        this.handleSwipe(this.direction);
      } else if (
        Math.abs(this.distY) >= this.threshold &&
        Math.abs(this.distX) <= this.restraint
      ) {
        this.handleSwipe(this.direction);
      }
    }
  }

  setRef(element) {
    this.surface = element;
  }

  handleSwipe(direction) {
    this.props.onSwipe(direction);
  }

  render() {
    return (
      <SwipeContainer
        onTouchStart={this.onTouchStart}
        onTouchEnd={this.onTouchEnd}
        innerRef={this.setRef}
      >
        {this.props.children}
      </SwipeContainer>
    );
  }
}

export default Swipe;
