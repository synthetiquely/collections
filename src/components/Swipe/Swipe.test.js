import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Swipe from './Swipe';

describe('Swipe Component', () => {
  it('should render and match its own snapshot', () => {
    const wrapper = mount(<Swipe />);
    const tree = renderer.create(<Swipe />).toJSON();

    expect(wrapper).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('should evoke onSwipe method passed as a props to the component', () => {
    const onSwipe = jest.fn();
    const wrapper = mount(<Swipe onSwipe={onSwipe} />);
    const direction = 'up';

    wrapper.instance().handleSwipe(direction);

    expect(onSwipe).toHaveBeenCalledWith(direction);
  });
});
