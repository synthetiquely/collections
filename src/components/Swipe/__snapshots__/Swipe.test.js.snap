// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swipe Component should render and match its own snapshot 1`] = `
<div
  className="css-zmgijg"
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
/>
`;
